In [1]:
import pandas as pd
import numpy as np

In [2]:
raw_data = pd.read_csv('data.csv')
raw_data.head()

register_index  post_id comment_id                author  \
0     145k12dpost  145k12d        NaN               SauloJr   
1  145hyeyjnlh1vx  145hyey    jnlh1vx  Illustrious_Risk3732   
2  1453u2ojnjtk7f  1453u2o    jnjtk7f       ProdigiousPlays   
3  1453u2ojnjtv9u  1453u2o    jnjtv9u                YJSubs   
4  1453u2ojnjusm9  1453u2o    jnjusm9           jamnewton22   

              datetime                                              title  \
0  2023-06-09 23:17:15  Reddit CEO Spez answered a grand total of 14 q...   
1  2023-06-09 23:42:43                                                NaN   
2  2023-06-09 16:47:51                                                NaN   
3  2023-06-09 16:49:52                                                NaN   
4  2023-06-09 16:55:57                                                NaN   

                                                 url  score  comments  \
0  https://www.reddit.com/r/technology/comments/1...    1.0       0.0   
1                                                NaN    1.0       NaN   
2                                                NaN    1.0       NaN   
3                                                NaN    1.0       NaN   
4                                                NaN    1.0       NaN   

                                                text  author_post_karma  \
0  Reddit CEO Spez answered a grand total of 14 q...            44092.0   
1  He even gives him self Awards to make sure he ...            17109.0   
2  RIF already announced it s shutting down at th...           118844.0   
3              I wonder which sub left on that day              50085.0   
4                  Oh I m sure this will go well lol           171705.0   

            tag  
0  Social Media  
1  Social Media  
2  Social Media  
3  Social Media  
4  Social Media

In [3]:
print(raw_data.shape)

(86916, 12)


In [4]:
raw_data.loc[raw_data['comment_id'].isna()]

register_index  post_id comment_id             author  \
0        145k12dpost  145k12d        NaN            SauloJr   
125      145g5xfpost  145g5xf        NaN     lily__brunette   
348      145f8kvpost  145f8kv        NaN  thebelsnickle1991   
431      1447ctspost  1447cts        NaN            ionedew   
455      145xmyxpost  145xmyx        NaN          dasun0218   
...              ...      ...        ...                ...   
86737    149fnx5post  149fnx5        NaN              hawlc   
86740    149fh87post  149fh87        NaN         RebootJobs   
86742    149eob7post  149eob7        NaN          OneOk2189   
86781    149egcvpost  149egcv        NaN         spsheridan   
86793    149drwspost  149drws        NaN    giuliomagnifico   

                  datetime                                              title  \
0      2023-06-09 23:17:15  Reddit CEO Spez answered a grand total of 14 q...   
125    2023-06-09 20:40:07  Reddit CEO Steve Huffman defends API changes i...   
348    2023-06-09 20:03:31  Netflix's password crackdown leads to massive ...   
431    2023-06-08 12:01:58  Sam Altman says he worries making ChatGPT was ...   
455    2023-06-10 11:17:50  16 Game-Changing AI Tools: Unleashing the Futu...   
...                    ...                                                ...   
86737  2023-06-14 18:46:14  Google risks forced breakup of ad business as ...   
86740  2023-06-14 18:38:56  Government Releases Unusually Frank Report on ...   
86742  2023-06-14 18:05:54  Paul Graham says remote work ‘does work initia...   
86781  2023-06-14 17:57:09  A team of physicists recently announced that t...   
86793  2023-06-14 17:30:04  Iran's 'quantum processor' turned out to be a ...   

                                                     url   score  comments  \
0      https://www.reddit.com/r/technology/comments/1...     1.0       0.0   
125    https://www.reddit.com/r/technology/comments/1...   859.0     232.0   
348    https://www.reddit.com/r/technology/comments/1...     0.0      14.0   
431    https://www.reddit.com/r/technology/comments/1...     3.0      10.0   
455    https://www.reddit.com/r/technology/comments/1...     3.0       0.0   
...                                                  ...     ...       ...   
86737  https://www.reddit.com/r/technology/comments/1...    20.0       2.0   
86740  https://www.reddit.com/r/technology/comments/1...     1.0       1.0   
86742  https://www.reddit.com/r/technology/comments/1...     0.0      39.0   
86781  https://www.reddit.com/r/technology/comments/1...    56.0      10.0   
86793  https://www.reddit.com/r/technology/comments/1...  1561.0     129.0   

                                                    text  author_post_karma  \
0      Reddit CEO Spez answered a grand total of 14 q...            44092.0   
125    Reddit CEO Steve Huffman defends API changes i...              542.0   
348    Netflix s password crackdown leads to massive ...           176771.0   
431    Sam Altman says he worries making ChatGPT was ...              466.0   
455    16 Game Changing AI Tools  Unleashing the Futu...              477.0   
...                                                  ...                ...   
86737  Google risks forced breakup of ad business as ...             3434.0   
86740  Government Releases Unusually Frank Report on ...            21171.0   
86742  Paul Graham says remote work  does work initia...             4627.0   
86781  A team of physicists recently announced that t...            17704.0   
86793  Iran s  quantum processor  turned out to be a ...            94361.0   

                           tag  
0                 Social Media  
125               Social Media  
348                   Business  
431    Artificial Intelligence  
455    Artificial Intelligence  
...                        ...  
86737                 Business  
86740                  Privacy  
86742                  Society  
86781       Nanotech/Materials  
86793                 Hardw

In [5]:
raw_data.loc[raw_data['text'].isna()]

Empty DataFrame
Columns: [register_index, post_id, comment_id, author, datetime, title, url, score, comments, text, author_post_karma, tag]
Index: []

# Problem statement
The primary goal of this analysis is to gain insights on whether sentiment score of a comment in a post can increase/decrease the post score.

In [6]:
# of posts and comments inside it
raw_data.groupby('post_id').count()['comment_id'].sort_values()

post_id
14a6xf2      0
1453sm0      0
1459yy6      0
145e9we      0
145k12d      0
          ... 
14bs8v6    790
14h1tdg    838
14g6hin    843
14ag85h    866
14b11yr    905
Name: comment_id, Length: 975, dtype: int64

In [7]:
raw_data.loc[raw_data['post_id']=='14b11yr']

register_index  post_id comment_id          author  \
7148   14b11yrjoe9sxp  14b11yr    joe9sxp  TorontoCorsair   
7149   14b11yrjoelsez  14b11yr    joelsez       olearyboy   
7150   14b11yrjodpccr  14b11yr    jodpccr     HombreMan24   
7151   14b11yrjodn1rb  14b11yr    jodn1rb         franker   
7152   14b11yrjoeefjx  14b11yr    joeefjx     Danteynero9   
...               ...      ...        ...             ...   
77314  14b11yrjofudap  14b11yr    jofudap           W3NTZ   
77315  14b11yrjog5cqk  14b11yr    jog5cqk       SomniumOv   
77316  14b11yrjofityh  14b11yr    jofityh      Baliverbes   
77317  14b11yrjof52d5  14b11yr    jof52d5     Good_ApoIIo   
77318  14b11yrjog22xs  14b11yr    jog22xs     Good_ApoIIo   

                  datetime title  url  score  comments  \
7148   2023-06-16 20:48:31   NaN  NaN    3.0       NaN   
7149   2023-06-16 22:30:07   NaN  NaN    3.0       NaN   
7150   2023-06-16 18:23:20   NaN  NaN    7.0       NaN   
7151   2023-06-16 18:07:00   NaN  NaN    5.0       NaN   
7152   2023-06-16 21:22:52   NaN  NaN    4.0       NaN   
...                    ...   ...  ...    ...       ...   
77314  2023-06-17 05:28:54   NaN  NaN    6.0       NaN   
77315  2023-06-17 07:59:15   NaN  NaN    6.0       NaN   
77316  2023-06-17 03:21:04   NaN  NaN    5.0       NaN   
77317  2023-06-17 01:13:02   NaN  NaN   56.0       NaN   
77318  2023-06-17 07:11:48   NaN  NaN    4.0       NaN   

                                                    text  author_post_karma  \
7148   First you get the higher mod priviledges  Then...             5564.0   
7149                                      r fuckyoukaren            23873.0   
7150   Can someone enlighten me in why a mod actually...             2473.0   
7151   DON T MAKE US HAVE TO STOP LETTING YOU WORK FO...            76064.0   
7152     What are you going to do  take away my salary              17317.0   
...                                                  ...                ...   
77314  My thought is if people come to reddit for hel...           293576.0   
77315    I heard on a podcast that they re undeleting...           171592.0   
77316  thankfully there s  r eyebleach  ah  shit  the...            44121.0   
77317  No don t mistake my meaning  They re still her...           251925.0   
77318  Gross  well there s our timeline it was only b...           251925.0   

                tag  
7148   Social Media  
7149   Social Media  
7150   Social Media  
7151   Social Media  
7152   Social Media  
...             ...  
77314  Social Media  
77315  Social Media  
77316  Social Media  
77317  Social Media  
77318  Social Media  

[906 rows x 12 columns]

In [8]:
# Example of a post with its commments
print('Number of comments',raw_data.loc[raw_data['post_id']=='14b11yr'].shape[0]-1)
print('Number of unique authors',raw_data.loc[raw_data['post_id']=='14b11yr']['author'].nunique())

Number of comments 905
Number of unique authors 654


## Find out whether a comment can be commented.

In [9]:
sum(raw_data.groupby('comment_id').count()['comments'].values)

0

## Filter posts without comments

In [10]:
raw_data.loc[(raw_data['comment_id'].isna()) & (raw_data['comments']>0)]

register_index  post_id comment_id             author  \
125      145g5xfpost  145g5xf        NaN     lily__brunette   
348      145f8kvpost  145f8kv        NaN  thebelsnickle1991   
431      1447ctspost  1447cts        NaN            ionedew   
456      145vqmzpost  145vqmz        NaN             Majnum   
509      1466uytpost  1466uyt        NaN             777fer   
...              ...      ...        ...                ...   
86737    149fnx5post  149fnx5        NaN              hawlc   
86740    149fh87post  149fh87        NaN         RebootJobs   
86742    149eob7post  149eob7        NaN          OneOk2189   
86781    149egcvpost  149egcv        NaN         spsheridan   
86793    149drwspost  149drws        NaN    giuliomagnifico   

                  datetime                                              title  \
125    2023-06-09 20:40:07  Reddit CEO Steve Huffman defends API changes i...   
348    2023-06-09 20:03:31  Netflix's password crackdown leads to massive ...   
431    2023-06-08 12:01:58  Sam Altman says he worries making ChatGPT was ...   
456    2023-06-10 09:31:26  Recent Mars Research Reveals New Discoveries a...   
509    2023-06-10 18:05:28  There's been a whopping 736 Tesla crashes invo...   
...                    ...                                                ...   
86737  2023-06-14 18:46:14  Google risks forced breakup of ad business as ...   
86740  2023-06-14 18:38:56  Government Releases Unusually Frank Report on ...   
86742  2023-06-14 18:05:54  Paul Graham says remote work ‘does work initia...   
86781  2023-06-14 17:57:09  A team of physicists recently announced that t...   
86793  2023-06-14 17:30:04  Iran's 'quantum processor' turned out to be a ...   

                                                     url   score  comments  \
125    https://www.reddit.com/r/technology/comments/1...   859.0     232.0   
348    https://www.reddit.com/r/technology/comments/1...     0.0      14.0   
431    https://www.reddit.com/r/technology/comments/1...     3.0      10.0   
456    https://www.reddit.com/r/technology/comments/1...    15.0       2.0   
509    https://www.reddit.com/r/technology/comments/1...    26.0      19.0   
...                                                  ...     ...       ...   
86737  https://www.reddit.com/r/technology/comments/1...    20.0       2.0   
86740  https://www.reddit.com/r/technology/comments/1...     1.0       1.0   
86742  https://www.reddit.com/r/technology/comments/1...     0.0      39.0   
86781  https://www.reddit.com/r/technology/comments/1...    56.0      10.0   
86793  https://www.reddit.com/r/technology/comments/1...  1561.0     129.0   

                                                    text  author_post_karma  \
125    Reddit CEO Steve Huffman defends API changes i...              542.0   
348    Netflix s password crackdown leads to massive ...           176771.0   
431    Sam Altman says he worries making ChatGPT was ...              466.0   
456    Recent Mars Research Reveals New Discoveries a...            17593.0   
509    There s been a whopping 736 Tesla crashes invo...              105.0   
...                                                  ...                ...   
86737  Google risks forced breakup of ad business as ...             3434.0   
86740  Government Releases Unusually Frank Report on ...            21171.0   
86742  Paul Graham says remote work  does work initia...             4627.0   
86781  A team of physicists recently announced that t...            17704.0   
86793  Iran s  quantum processor  turned out to be a ...            94361.0   

                           tag  
125               Social Media  
348                   Business  
431    Artificial Intelligence  
456                      Space  
509             Transportation  
...                        ...  
86737                 Business  
86740                  Privacy  
86742                  Society  
86781       Nanotech/Materials  
86793                 Hardw

The total number of posts is 975. Total number of posts with at least 1 comment is 909. There are 66 posts without comments. We can exclude them.

In [11]:
posts = raw_data.loc[(raw_data['comment_id'].isna()) & (raw_data['comments']>0)]
posts.shape

(909, 12)

In [12]:
posts.isna().sum()

register_index         0
post_id                0
comment_id           909
author                 0
datetime               0
title                  0
url                    0
score                  0
comments               0
text                   0
author_post_karma     12
tag                    0
dtype: int64

In [13]:
comments = raw_data.loc[(raw_data['post_id'].isin(set(posts['post_id'])))&(~raw_data['comment_id'].isna())]
comments.head()

register_index  post_id comment_id                author  \
1  145hyeyjnlh1vx  145hyey    jnlh1vx  Illustrious_Risk3732   
2  1453u2ojnjtk7f  1453u2o    jnjtk7f       ProdigiousPlays   
3  1453u2ojnjtv9u  1453u2o    jnjtv9u                YJSubs   
4  1453u2ojnjusm9  1453u2o    jnjusm9           jamnewton22   
5  1453u2ojnjvwlv  1453u2o    jnjvwlv            TriLink710   

              datetime title  url  score  comments  \
1  2023-06-09 23:42:43   NaN  NaN    1.0       NaN   
2  2023-06-09 16:47:51   NaN  NaN    1.0       NaN   
3  2023-06-09 16:49:52   NaN  NaN    1.0       NaN   
4  2023-06-09 16:55:57   NaN  NaN    1.0       NaN   
5  2023-06-09 17:03:15   NaN  NaN    1.0       NaN   

                                                text  author_post_karma  \
1  He even gives him self Awards to make sure he ...            17109.0   
2  RIF already announced it s shutting down at th...           118844.0   
3              I wonder which sub left on that day              50085.0   
4                  Oh I m sure this will go well lol           171705.0   
5  Oh boy i can t wait for him to say  the devs a...            70192.0   

            tag  
1  Social Media  
2  Social Media  
3  Social Media  
4  Social Media  
5  Social Media

In [14]:
comments=comments.drop(['register_index','title','url','comments'], axis=1)
comments.head()

post_id comment_id                author             datetime  score  \
1  145hyey    jnlh1vx  Illustrious_Risk3732  2023-06-09 23:42:43    1.0   
2  1453u2o    jnjtk7f       ProdigiousPlays  2023-06-09 16:47:51    1.0   
3  1453u2o    jnjtv9u                YJSubs  2023-06-09 16:49:52    1.0   
4  1453u2o    jnjusm9           jamnewton22  2023-06-09 16:55:57    1.0   
5  1453u2o    jnjvwlv            TriLink710  2023-06-09 17:03:15    1.0   

                                                text  author_post_karma  \
1  He even gives him self Awards to make sure he ...            17109.0   
2  RIF already announced it s shutting down at th...           118844.0   
3              I wonder which sub left on that day              50085.0   
4                  Oh I m sure this will go well lol           171705.0   
5  Oh boy i can t wait for him to say  the devs a...            70192.0   

            tag  
1  Social Media  
2  Social Media  
3  Social Media  
4  Social Media  
5  Social Media

In [15]:
posts = posts.drop('comment_id', axis=1)
posts.head()

register_index  post_id             author             datetime  \
125    145g5xfpost  145g5xf     lily__brunette  2023-06-09 20:40:07   
348    145f8kvpost  145f8kv  thebelsnickle1991  2023-06-09 20:03:31   
431    1447ctspost  1447cts            ionedew  2023-06-08 12:01:58   
456    145vqmzpost  145vqmz             Majnum  2023-06-10 09:31:26   
509    1466uytpost  1466uyt             777fer  2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  score  comments  \
125  https://www.reddit.com/r/technology/comments/1...  859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...    0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...    3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...   15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...   26.0      19.0   

                                                  text  author_post_karma  \
125  Reddit CEO Steve Huffman defends API changes i...              542.0   
348  Netflix s password crackdown leads to massive ...           176771.0   
431  Sam Altman says he worries making ChatGPT was ...              466.0   
456  Recent Mars Research Reveals New Discoveries a...            17593.0   
509  There s been a whopping 736 Tesla crashes invo...              105.0   

                         tag  
125             Social Media  
348                 Business  
431  Artificial Intelligence  
456                    Space  
509           Transportation

### Separating columns for comments and posts

In [16]:
cols_comment = comments.columns.to_list()
cols_comment = cols_comment[2:]
col_dict = dict.fromkeys(cols_comment)
for i in cols_comment:
    col_dict[i] = 'comment_'+i
col_dict

{'author': 'comment_author',
 'datetime': 'comment_datetime',
 'score': 'comment_score',
 'text': 'comment_text',
 'author_post_karma': 'comment_author_post_karma',
 'tag': 'comment_tag'}

In [17]:
comments.rename(columns=col_dict, inplace=True)
comments.head()

post_id comment_id        comment_author     comment_datetime  \
1  145hyey    jnlh1vx  Illustrious_Risk3732  2023-06-09 23:42:43   
2  1453u2o    jnjtk7f       ProdigiousPlays  2023-06-09 16:47:51   
3  1453u2o    jnjtv9u                YJSubs  2023-06-09 16:49:52   
4  1453u2o    jnjusm9           jamnewton22  2023-06-09 16:55:57   
5  1453u2o    jnjvwlv            TriLink710  2023-06-09 17:03:15   

   comment_score                                       comment_text  \
1            1.0  He even gives him self Awards to make sure he ...   
2            1.0  RIF already announced it s shutting down at th...   
3            1.0              I wonder which sub left on that day     
4            1.0                  Oh I m sure this will go well lol   
5            1.0  Oh boy i can t wait for him to say  the devs a...   

   comment_author_post_karma   comment_tag  
1                    17109.0  Social Media  
2                   118844.0  Social Media  
3                    50085.0  Social Media  
4                   171705.0  Social Media  
5                    70192.0  Social Media

In [18]:
posts.drop(['register_index', 'text'], axis=1, inplace=True)

In [19]:
posts.head()

post_id             author             datetime  \
125  145g5xf     lily__brunette  2023-06-09 20:40:07   
348  145f8kv  thebelsnickle1991  2023-06-09 20:03:31   
431  1447cts            ionedew  2023-06-08 12:01:58   
456  145vqmz             Majnum  2023-06-10 09:31:26   
509  1466uyt             777fer  2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  score  comments  \
125  https://www.reddit.com/r/technology/comments/1...  859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...    0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...    3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...   15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...   26.0      19.0   

     author_post_karma                      tag  
125              542.0             Social Media  
348           176771.0                 Business  
431              466.0  Artificial Intelligence  
456            17593.0                    Space  
509              105.0           Transportation

In [20]:
cols_post = ['author', 'datetime', 'score', 'tag']
col_dict = dict.fromkeys(cols_post)
for i in cols_post:
    col_dict[i] = 'post_'+i
col_dict

{'author': 'post_author',
 'datetime': 'post_datetime',
 'score': 'post_score',
 'tag': 'post_tag'}

In [21]:
posts.rename(columns=col_dict, inplace=True)
posts.head()

post_id        post_author        post_datetime  \
125  145g5xf     lily__brunette  2023-06-09 20:40:07   
348  145f8kv  thebelsnickle1991  2023-06-09 20:03:31   
431  1447cts            ionedew  2023-06-08 12:01:58   
456  145vqmz             Majnum  2023-06-10 09:31:26   
509  1466uyt             777fer  2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  post_score  comments  \
125  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...         0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...         3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...        15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...        26.0      19.0   

     author_post_karma                 post_tag  
125              542.0             Social Media  
348           176771.0                 Business  
431              466.0  Artificial Intelligence  
456            17593.0                    Space  
509              105.0           Transportation

### Merging data

In [28]:
data = posts.merge(comments, on='post_id', how='left')
print(data.shape)
data.head()

(85953, 16)


post_id     post_author        post_datetime  \
0  145g5xf  lily__brunette  2023-06-09 20:40:07   
1  145g5xf  lily__brunette  2023-06-09 20:40:07   
2  145g5xf  lily__brunette  2023-06-09 20:40:07   
3  145g5xf  lily__brunette  2023-06-09 20:40:07   
4  145g5xf  lily__brunette  2023-06-09 20:40:07   

                                               title  \
0  Reddit CEO Steve Huffman defends API changes i...   
1  Reddit CEO Steve Huffman defends API changes i...   
2  Reddit CEO Steve Huffman defends API changes i...   
3  Reddit CEO Steve Huffman defends API changes i...   
4  Reddit CEO Steve Huffman defends API changes i...   

                                                 url  post_score  comments  \
0  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
1  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
2  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
3  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
4  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   

   author_post_karma      post_tag comment_id     comment_author  \
0              542.0  Social Media    jnkvhp0         lunarmedic   
1              542.0  Social Media    jnkvmqh        CL-MotoTech   
2              542.0  Social Media    jnkti6m  whoneedsusernames   
3              542.0  Social Media    jnkyr2b         celtic1888   
4              542.0  Social Media    jnl5w5w    InternetArtisan   

      comment_datetime  comment_score  \
0  2023-06-09 21:01:30          484.0   
1  2023-06-09 21:02:30          378.0   
2  2023-06-09 20:47:55          172.0   
3  2023-06-09 21:24:12          125.0   
4  2023-06-09 22:15:35           19.0   

                                        comment_text  \
0   Direct link to AMA  https   www reddit com r ...   
1  The last I checked he had 14 responses total  ...   
2                  Incorrect title    poorly defends   
3  He ran off to his bunker after spouting nonsen...   
4   One of his most telling answers came in respo...   

   comment_author_post_karma   comment_tag  
0                    15835.0  Social Media  
1                    46444.0  Social Media  
2                    82843.0  Social Media  
3                   592646.0  Social Media  
4                   119550.0  Social Media

### Checking whether post tag is same with comment tag

In [29]:
data.loc[~data['post_tag'].eq(data['comment_tag'])]

post_id          post_author        post_datetime  \
5775   141fc0m        Stiven_Crysis  2023-06-05 13:54:38   
13987  1436x6l           Saltedline  2023-06-07 07:57:32   
15269  149dnod           whatsalamp  2023-06-14 17:25:12   
24243  145mowf             Ssider69  2023-06-10 01:18:42   
37444  14jfvaj            100GB-CSV  2023-06-26 12:22:30   
45477  14h0qny          TommyAdagio  2023-06-23 14:57:49   
49613  14g5evf               Majnum  2023-06-22 14:43:34   
53200  14fd99k               konorM  2023-06-21 17:03:02   
58978  14eagk6     OregonTripleBeam  2023-06-20 12:57:30   
74080  14b0az1      1_Portgas_D_Ace  2023-06-16 15:48:56   
82578  14a3ia3  ChickenTeriyakiBoy1  2023-06-15 14:13:16   
85440  149jhb8          ethereal3xp  2023-06-14 21:18:57   

                                                   title  \
5775   With PowerVia, Intel Achieves a Chipmaking Bre...   
13987  Supercomputer Fugaku to help develop Japanese ...   
15269  An Anti-Porn App Put Him in Jail and His Famil...   
24243               Engine treatment can slash jet noise   
37444  WSJ News Exclusive | Databricks Strikes $1.3 B...   
45477  Gearlinx emerges from stealth to deliver ‘Netw...   
49613  Satellite-based laser communication: A rival t...   
53200  Drones take to the waves: Saildrones are getti...   
58978  Portland's emergency communications bureau lau...   
74080  NASA agreements to support work on commercial ...   
82578  A Shady Chinese Firm’s Encryption Chips Got In...   
85440  Broadcom squeezed Samsung, now South Korea’s s...   

                                                     url  post_score  \
5775   https://www.reddit.com/r/technology/comments/1...        14.0   
13987  https://www.reddit.com/r/technology/comments/1...        13.0   
15269  https://www.reddit.com/r/technology/comments/1...         0.0   
24243  https://www.reddit.com/r/technology/comments/1...        36.0   
37444  https://www.reddit.com/r/technology/comments/1...        10.0   
45477  https://www.reddit.com/r/technology/comments/1...         0.0   
49613  https://www.reddit.com/r/technology/comments/1...         2.0   
53200  https://www.reddit.com/r/technology/comments/1...        32.0   
58978  https://www.reddit.com/r/technology/comments/1...        10.0   
74080  https://www.reddit.com/r/technology/comments/1...         6.0   
82578  https://www.reddit.com/r/technology/comments/1...        14.0   
85440  https://www.reddit.com/r/technology/comments/1...         7.0   

       comments  author_post_karma                 post_tag comment_id  \
5775        1.0             1357.0                 Hardware        NaN   
13987       1.0            27471.0  Artificial Intelligence        NaN   
15269       1.0             3107.0                  Privacy        NaN   
24243       1.0           322400.0           Transportation        NaN   
37444       1.0              100.0         Machine Learning        NaN   
45477       1.0             5198.0       Networking/Telecom        NaN   
49613       1.0            17654.0                 Hardware        NaN   
53200       1.0            78233.0      Robotics/Automation        NaN   
58978       1.0           257290.0  Artificial Intelligence        NaN   
74080       1.0              231.0                    Space        NaN   
82578       1.0            16752.0                 Security        NaN   
85440       1.0            32522.0       Nanotech/Materials        NaN   

      comment_author comment_datetime  comment_score comment_text  \
5775             NaN              NaN            NaN          NaN   
13987            NaN              NaN            NaN          NaN   
15269            NaN              NaN            NaN          NaN   
24243            NaN              NaN            NaN          NaN   
37444            NaN              NaN            NaN          NaN   
45477            NaN              NaN            NaN          NaN   
49613            NaN              NaN            NaN          

Above code also finds out the posts that have comments but it does not exist in the data. So let's remove them.

In [31]:
data = data.loc[~data['comment_text'].isna()]
print(data.shape)
data.head()

post_id     post_author        post_datetime  \
0  145g5xf  lily__brunette  2023-06-09 20:40:07   
1  145g5xf  lily__brunette  2023-06-09 20:40:07   
2  145g5xf  lily__brunette  2023-06-09 20:40:07   
3  145g5xf  lily__brunette  2023-06-09 20:40:07   
4  145g5xf  lily__brunette  2023-06-09 20:40:07   

                                               title  \
0  Reddit CEO Steve Huffman defends API changes i...   
1  Reddit CEO Steve Huffman defends API changes i...   
2  Reddit CEO Steve Huffman defends API changes i...   
3  Reddit CEO Steve Huffman defends API changes i...   
4  Reddit CEO Steve Huffman defends API changes i...   

                                                 url  post_score  comments  \
0  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
1  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
2  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
3  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
4  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   

   author_post_karma      post_tag comment_id     comment_author  \
0              542.0  Social Media    jnkvhp0         lunarmedic   
1              542.0  Social Media    jnkvmqh        CL-MotoTech   
2              542.0  Social Media    jnkti6m  whoneedsusernames   
3              542.0  Social Media    jnkyr2b         celtic1888   
4              542.0  Social Media    jnl5w5w    InternetArtisan   

      comment_datetime  comment_score  \
0  2023-06-09 21:01:30          484.0   
1  2023-06-09 21:02:30          378.0   
2  2023-06-09 20:47:55          172.0   
3  2023-06-09 21:24:12          125.0   
4  2023-06-09 22:15:35           19.0   

                                        comment_text  \
0   Direct link to AMA  https   www reddit com r ...   
1  The last I checked he had 14 responses total  ...   
2                  Incorrect title    poorly defends   
3  He ran off to his bunker after spouting nonsen...   
4   One of his most telling answers came in respo...   

   comment_author_post_karma   comment_tag  
0                    15835.0  Social Media  
1                    46444.0  Social Media  
2                    82843.0  Social Media  
3                   592646.0  Social Media  
4                   119550.0  Social Media

## Finding out invalid data

In [32]:
data.loc[data['comment_author'].isin(set(data['post_author']))]

post_id      post_author        post_datetime  \
21     145g5xf   lily__brunette  2023-06-09 20:40:07   
22     145g5xf   lily__brunette  2023-06-09 20:40:07   
30     145g5xf   lily__brunette  2023-06-09 20:40:07   
51     145g5xf   lily__brunette  2023-06-09 20:40:07   
61     145g5xf   lily__brunette  2023-06-09 20:40:07   
...        ...              ...                  ...   
85806  149eob7        OneOk2189  2023-06-14 18:05:54   
85816  149egcv       spsheridan  2023-06-14 17:57:09   
85859  149drws  giuliomagnifico  2023-06-14 17:30:04   
85905  149drws  giuliomagnifico  2023-06-14 17:30:04   
85913  149drws  giuliomagnifico  2023-06-14 17:30:04   

                                                   title  \
21     Reddit CEO Steve Huffman defends API changes i...   
22     Reddit CEO Steve Huffman defends API changes i...   
30     Reddit CEO Steve Huffman defends API changes i...   
51     Reddit CEO Steve Huffman defends API changes i...   
61     Reddit CEO Steve Huffman defends API changes i...   
...                                                  ...   
85806  Paul Graham says remote work ‘does work initia...   
85816  A team of physicists recently announced that t...   
85859  Iran's 'quantum processor' turned out to be a ...   
85905  Iran's 'quantum processor' turned out to be a ...   
85913  Iran's 'quantum processor' turned out to be a ...   

                                                     url  post_score  \
21     https://www.reddit.com/r/technology/comments/1...       859.0   
22     https://www.reddit.com/r/technology/comments/1...       859.0   
30     https://www.reddit.com/r/technology/comments/1...       859.0   
51     https://www.reddit.com/r/technology/comments/1...       859.0   
61     https://www.reddit.com/r/technology/comments/1...       859.0   
...                                                  ...         ...   
85806  https://www.reddit.com/r/technology/comments/1...         0.0   
85816  https://www.reddit.com/r/technology/comments/1...        56.0   
85859  https://www.reddit.com/r/technology/comments/1...      1561.0   
85905  https://www.reddit.com/r/technology/comments/1...      1561.0   
85913  https://www.reddit.com/r/technology/comments/1...      1561.0   

       comments  author_post_karma            post_tag comment_id  \
21        232.0              542.0        Social Media    jnl4ayk   
22        232.0              542.0        Social Media    jnlawh9   
30        232.0              542.0        Social Media    jnl48nw   
51        232.0              542.0        Social Media    jnl1pfs   
61        232.0              542.0        Social Media    jnkxku3   
...         ...                ...                 ...        ...   
85806      39.0             4627.0             Society    jo5e0s5   
85816      10.0            17704.0  Nanotech/Materials    jo5hw9o   
85859     129.0            94361.0            Hardware    jo75xm0   
85905     129.0            94361.0            Hardware    jo624yz   
85913     129.0            94361.0            Hardware    jo8272u   

         comment_author     comment_datetime  comment_score  \
21              ledim35  2023-06-09 22:03:37            3.0   
22        Crack_uv_N0on  2023-06-09 22:54:07            3.0   
30              ledim35  2023-06-09 22:03:13            1.0   
51               oDDmON  2023-06-09 21:44:54           69.0   
61     LittleRickyPemba  2023-06-09 21:15:58           80.0   
...                 ...                  ...            ...   
85806        bitfriend6  2023-06-14 20:58:03            3.0   
85816         [deleted]  2023-06-14 21:26:04            3.0   
85859         [deleted]  2023-06-15 06:47:23            0.0   
85905         [deleted]  2023-06-15 00:07:16            7.0   
85913         [deleted]  2023-06-15 13:23:47            1.0   

                                            comment_text  \
21                                He s just being silly    
22     The beatings will continue until mo

### Check `deleted` or `removed` comments

In [50]:
for i in set(data['comment_author']):
    if '[deleted]' in i:
        print(i)

[deleted]


In [59]:
data.loc[data['post_id']=='149drws']

post_id      post_author        post_datetime  \
85827  149drws  giuliomagnifico  2023-06-14 17:30:04   
85828  149drws  giuliomagnifico  2023-06-14 17:30:04   
85829  149drws  giuliomagnifico  2023-06-14 17:30:04   
85830  149drws  giuliomagnifico  2023-06-14 17:30:04   
85831  149drws  giuliomagnifico  2023-06-14 17:30:04   
...        ...              ...                  ...   
85948  149drws  giuliomagnifico  2023-06-14 17:30:04   
85949  149drws  giuliomagnifico  2023-06-14 17:30:04   
85950  149drws  giuliomagnifico  2023-06-14 17:30:04   
85951  149drws  giuliomagnifico  2023-06-14 17:30:04   
85952  149drws  giuliomagnifico  2023-06-14 17:30:04   

                                                   title  \
85827  Iran's 'quantum processor' turned out to be a ...   
85828  Iran's 'quantum processor' turned out to be a ...   
85829  Iran's 'quantum processor' turned out to be a ...   
85830  Iran's 'quantum processor' turned out to be a ...   
85831  Iran's 'quantum processor' turned out to be a ...   
...                                                  ...   
85948  Iran's 'quantum processor' turned out to be a ...   
85949  Iran's 'quantum processor' turned out to be a ...   
85950  Iran's 'quantum processor' turned out to be a ...   
85951  Iran's 'quantum processor' turned out to be a ...   
85952  Iran's 'quantum processor' turned out to be a ...   

                                                     url  post_score  \
85827  https://www.reddit.com/r/technology/comments/1...      1561.0   
85828  https://www.reddit.com/r/technology/comments/1...      1561.0   
85829  https://www.reddit.com/r/technology/comments/1...      1561.0   
85830  https://www.reddit.com/r/technology/comments/1...      1561.0   
85831  https://www.reddit.com/r/technology/comments/1...      1561.0   
...                                                  ...         ...   
85948  https://www.reddit.com/r/technology/comments/1...      1561.0   
85949  https://www.reddit.com/r/technology/comments/1...      1561.0   
85950  https://www.reddit.com/r/technology/comments/1...      1561.0   
85951  https://www.reddit.com/r/technology/comments/1...      1561.0   
85952  https://www.reddit.com/r/technology/comments/1...      1561.0   

       comments  author_post_karma  post_tag comment_id       comment_author  \
85827     129.0            94361.0  Hardware    jo5psts         CrunchySkull   
85828     129.0            94361.0  Hardware    jo5bmbv      muff_nose_diver   
85829     129.0            94361.0  Hardware    jo6v6wn     vortex_unlimited   
85830     129.0            94361.0  Hardware    jo88h2a  TheGreat23Recession   
85831     129.0            94361.0  Hardware    jo4z6b3    CryptoNerdSmacker   
...         ...                ...       ...        ...                  ...   
85948     129.0            94361.0  Hardware    jo7dx9v         talonredwing   
85949     129.0            94361.0  Hardware    jo77nni            MrCamlost   
85950     129.0            94361.0  Hardware    jo5i09h               a4mula   
85951     129.0            94361.0  Hardware    jo7eblf                ninta   
85952     129.0            94361.0  Hardware    jo5izf2              Sorge74   

          comment_datetime  comment_score  \
85827  2023-06-14 22:26:22            2.0   
85828  2023-06-14 20:41:18            2.0   
85829  2023-06-15 04:33:23            2.0   
85830  2023-06-15 14:13:33            1.0   
85831  2023-06-14 19:14:16          391.0   
...                    ...            ...   
85948  2023-06-15 08:45:37            2.0   
85949  2023-06-15 07:11:44            0.0   
85950  2023-06-14 21:26:52            2.0   
85951  2023-06-15 08:51:38            4.0   
85952  2023-06-14 21:34:09            1.0   

                                            comment_text  \
85827  Iran and quantum anything in the same sentence...   
85828     Look at their little outfits  They re so cute    
85829    I wouldn t want to piss off Iranian official...   
85830 

In [72]:
data['comment_text']=data['comment_text'].astype(str)
data['comment_text']=data['comment_text'].str.strip()

In [78]:
data.loc[data['comment_text'].isin(['deleted','removed'])]

post_id        post_author        post_datetime  \
307    149gn6p  thebelsnickle1991  2023-06-14 19:25:04   
405    14a3lyn             777fer  2023-06-15 14:17:39   
413    14a3lyn             777fer  2023-06-15 14:17:39   
420    14a3lyn             777fer  2023-06-15 14:17:39   
441    149ipec             777fer  2023-06-14 20:47:59   
...        ...                ...                  ...   
85705  149h82b        return2ozma  2023-06-14 19:48:44   
85745  149h82b        return2ozma  2023-06-14 19:48:44   
85816  149egcv         spsheridan  2023-06-14 17:57:09   
85905  149drws    giuliomagnifico  2023-06-14 17:30:04   
85913  149drws    giuliomagnifico  2023-06-14 17:30:04   

                                                   title  \
307    Reddit CEO assures employees that API pricing ...   
405    Some doctors are using AI-chatbots like ChatGP...   
413    Some doctors are using AI-chatbots like ChatGP...   
420    Some doctors are using AI-chatbots like ChatGP...   
441    The 14-year-old hired by Elon Musk's SpaceX is...   
...                                                  ...   
85705  The Reddit Protest Is a Battle for the Soul of...   
85745  The Reddit Protest Is a Battle for the Soul of...   
85816  A team of physicists recently announced that t...   
85905  Iran's 'quantum processor' turned out to be a ...   
85913  Iran's 'quantum processor' turned out to be a ...   

                                                     url  post_score  \
307    https://www.reddit.com/r/technology/comments/1...        12.0   
405    https://www.reddit.com/r/technology/comments/1...        36.0   
413    https://www.reddit.com/r/technology/comments/1...        36.0   
420    https://www.reddit.com/r/technology/comments/1...        36.0   
441    https://www.reddit.com/r/technology/comments/1...        64.0   
...                                                  ...         ...   
85705  https://www.reddit.com/r/technology/comments/1...       358.0   
85745  https://www.reddit.com/r/technology/comments/1...       358.0   
85816  https://www.reddit.com/r/technology/comments/1...        56.0   
85905  https://www.reddit.com/r/technology/comments/1...      1561.0   
85913  https://www.reddit.com/r/technology/comments/1...      1561.0   

       comments  author_post_karma                 post_tag comment_id  \
307        17.0           177515.0                 Business    jo5dhvd   
405        35.0               99.0  Artificial Intelligence    jo9w0nn   
413        35.0               99.0  Artificial Intelligence    joa07q5   
420        35.0               99.0  Artificial Intelligence    jo9i7tu   
441        48.0               99.0                  Society    jo5pb67   
...         ...                ...                      ...        ...   
85705     269.0           521621.0             Social Media    jo5sh1v   
85745     269.0           521621.0             Social Media    jo5e8ip   
85816      10.0            17704.0       Nanotech/Materials    jo5hw9o   
85905     129.0            94361.0                 Hardware    jo624yz   
85913     129.0            94361.0                 Hardware    jo8272u   

      comment_author     comment_datetime  comment_score comment_text  \
307        [deleted]  2023-06-14 20:54:22            1.0      deleted   
405        [deleted]  2023-06-15 21:23:13            0.0      removed   
413        [deleted]  2023-06-15 21:53:30           -1.0      removed   
420        [deleted]  2023-06-15 19:47:41            1.0      deleted   
441        [deleted]  2023-06-14 22:22:29          -55.0      removed   
...              ...                  ...            ...          ...   
85705      [deleted]  2023-06-14 22:47:50           -4.0      deleted   
85745      [deleted]  2023-06-14 20:59:31           -2.0      deleted   
85816      [deleted]  2023-06-14 21:26:04            3.0      deleted   
85905      [deleted]  2023-06-15 00:07:16            7.0      deleted   
85913      [deleted]  2023-06-15